libraries/standard/mqtt/integration-test/mqtt_system_utest.c

```diff
@@ -0,0 +1,700 @@
+}
+
+/* ============================   UNITY FIXTURES ============================ */
+
+/* Called before each test method. */
+void setUp()
+{
+    /* Reset file-scoped global variables. */
+    receivedSubAck = false;
+    receivedUnsubAck = false;
+    receivedPubAck = false;
+    receivedPubRec = false;
+    receivedPubRel = false;
+    receivedPubComp = false;
+    persistentSession = false;
+    memset( &incomingInfo, 0u, sizeof( MQTTPublishInfo_t ) );
+
+    /* Establish a TCP connection with the server endpoint. */
+    TEST_ASSERT_EQUAL( EXIT_SUCCESS,
+                       tcpConnectToServer( BROKER_ENDPOINT, BROKER_PORT, &tcpSocket ) );
+    TEST_ASSERT_NOT_EQUAL( -1, tcpSocket );
+
+    /* Establish TLS connection on top of TCP connection. */
+    TEST_ASSERT_EQUAL( EXIT_SUCCESS, tlsSetup( tcpSocket, SERVER_ROOT_CA_CERT_PATH, &pSslContext ) );
+    TEST_ASSERT_NOT_NULL( pSslContext );
+
+    /* Establish MQTT session on top of the TCP+TLS connection. */
+    establishMqttSession( &context, pSslContext, true, &persistentSession );
+}
+
+/* Called after each test method. */
+void tearDown()
+{
+    /* Free memory, if allocated during test case execution. */
+    if( incomingInfo.pTopicName != NULL )
+    {
+        free( ( void * ) incomingInfo.pTopicName );
+    }
+
+    if( incomingInfo.pPayload != NULL )
+    {
+        free( ( void * ) incomingInfo.pPayload );
+    }
+
+    /* Terminate MQTT connection. */
+    TEST_ASSERT_EQUAL( MQTTSuccess, MQTT_Disconnect( &context ) );
+
+    /* Terminate TLS session. */
+    ( void ) closeTlsSession( pSslContext );
+
+    /* Terminate TCP connection. */
+    ( void ) closeTcpConnection( tcpSocket );
+}
+
+/* ========================== Test Cases ============================ */
+
+/**
+ * @brief Tests Subscribe and Publish operations with the MQTT broken using QoS 0.
+ * The test subscribes to a topic, and then publishes to the same topic. The
+ * broker is expected to route the publish message back to the test.
+ */
+void test_MQTT_Subscribe_Publish_With_Qos_0( void )
+{
+    /* Subscribe to a topic with Qos 0. */
+    TEST_ASSERT_EQUAL( MQTTSuccess, subscribeToTopic(
+                           &context, TEST_MQTT_TOPIC, MQTTQoS0 ) );
+
+    /* We do not expect a SUBACK from the broker for the subscribe operation. */
+    TEST_ASSERT_EQUAL( MQTTSuccess,
+                       MQTT_ProcessLoop( &context, MQTT_PROCESS_LOOP_TIMEOUT_MS ) );
+
+    /* Publish to the same topic, that we subscribed to, with Qos 1. */
+    TEST_ASSERT_EQUAL( MQTTSuccess, publishToTopic(
+                           &context, TEST_MQTT_TOPIC, MQTTQoS0 ) );
+
+    /* Call the MQTT library for the expectation to read an incoming PUBLISH for
+     * the same message that we published (as we have subscribed to the same topic). */
+    TEST_ASSERT_EQUAL( MQTTSuccess,
+                       MQTT_ProcessLoop( &context, MQTT_PROCESS_LOOP_TIMEOUT_MS ) );
+    /* We do not expect a PUBACK from the broker for the PUBLISH. */
```

Comment:
We do not expect a PUBACK from the broker for the **QoS0** PUBLISH.

Reply:
Fixed